board/safety/safety_ford.h

```diff
@@ -1,112 +1,219 @@
-//      cancel button
-//      accel rising edge
-//      brake rising edge
-//      brake > 0mph
+const struct lookup_t FORD_LOOKUP_ANGLE_RATE_UP = {
+  {2., 7., 17.},
+  {5., .8, .25}};
+
+const struct lookup_t FORD_LOOKUP_ANGLE_RATE_DOWN = {
+  {2., 7., 17.},
+  {5., 3.5, .8}};
+
+// Signal: LatCtlPath_An_Actl
+// Factor: 0.0005
+const int FORD_DEG_TO_CAN = 2000;
+
+// Safety-relevant CAN messages for Ford vehicles.
+#define MSG_STEERING_PINION_DATA      0x07E  // RX from PSCM, for steering pinion angle
+#define MSG_ENG_BRAKE_DATA            0x165  // RX from PCM, for driver brake pedal and cruise state
+#define MSG_ENG_VEHICLE_SP_THROTTLE2  0x202  // RX from PCM, for vehicle speed
+#define MSG_ENG_VEHICLE_SP_THROTTLE   0x204  // RX from PCM, for driver throttle input
+#define MSG_STEERING_DATA_FD1         0x083  // TX by OP, ACC control buttons for cancel
+#define MSG_LANE_ASSIST_DATA1         0x3CA  // TX by OP, Lane Keeping Assist
+#define MSG_LATERAL_MOTION_CONTROL    0x3D3  // TX by OP, Lane Centering Assist
+#define MSG_IPMA_DATA                 0x3D8  // TX by OP, IPMA HUD user interface
+
+// CAN bus numbers.
+#define FORD_MAIN 0U
+#define FORD_CAM  2U
+
+const CanMsg FORD_TX_MSGS[] = {
+  {MSG_STEERING_DATA_FD1, 0, 8},
+  {MSG_LANE_ASSIST_DATA1, 0, 8},
+  {MSG_LATERAL_MOTION_CONTROL, 0, 8},
+  {MSG_IPMA_DATA, 0, 8},
+};
+#define FORD_TX_LEN (sizeof(FORD_TX_MSGS) / sizeof(FORD_TX_MSGS[0]))
+
+AddrCheckStruct ford_addr_checks[] = {
+  {.msg = {{MSG_STEERING_PINION_DATA, 0, 8, .expected_timestep = 10000U}, { 0 }, { 0 }}},
+  {.msg = {{MSG_ENG_BRAKE_DATA, 0, 8, .expected_timestep = 100000U}, { 0 }, { 0 }}},
+  {.msg = {{MSG_ENG_VEHICLE_SP_THROTTLE2, 0, 8, .expected_timestep = 20000U}, { 0 }, { 0 }}},
+  {.msg = {{MSG_ENG_VEHICLE_SP_THROTTLE, 0, 8, .expected_timestep = 10000U}, { 0 }, { 0 }}},
+};
+#define FORD_ADDR_CHECK_LEN (sizeof(ford_addr_checks) / sizeof(ford_addr_checks[0]))
+addr_checks ford_rx_checks = {ford_addr_checks, FORD_ADDR_CHECK_LEN};
+
+
+static const addr_checks* ford_init(int16_t param) {
+  UNUSED(param);
 
+  controls_allowed = false;
+  relay_malfunction_reset();
+
+  return &ford_rx_checks;
+}
 
 static int ford_rx_hook(CANPacket_t *to_push) {
+  bool valid = addr_safety_check(to_push, &ford_rx_checks, NULL, NULL, NULL);
 
-  int addr = GET_ADDR(to_push);
-  int bus = GET_BUS(to_push);
-  bool unsafe_allow_gas = unsafe_mode & UNSAFE_DISABLE_DISENGAGE_ON_GAS;
+  if (valid && (GET_BUS(to_push) == FORD_MAIN)) {
+    int addr = GET_ADDR(to_push);
 
-  if (addr == 0x217) {
-    // wheel speeds are 14 bits every 16
-    vehicle_moving = false;
-    for (int i = 0; i < 8; i += 2) {
-      vehicle_moving |= GET_BYTE(to_push, i) | (GET_BYTE(to_push, (int)(i + 1)) & 0xFCU);
+    // Update steering angle
+    if (addr == MSG_STEERING_PINION_DATA) {
+      // Steering angle: (0.1 * val) - 3200 in deg
+      // Signal: StePinRelInit_An_Sns
+      int angle_meas_new = ((GET_BYTE(to_push, 0) << 8) | GET_BYTE(to_push, 1)) - 3200U;
```

Comment:
Need your scaling in addition to your offset, do you not? Haven't checked to see how this interplays with your steering command being in radians; I haven't used angle control before.

```suggestion
      // Steering angle: (0.1 * val) - 3200 in deg
      // Signal: StePinRelInit_An_Sns
      int angle_meas_new = ((GET_BYTE(to_push, 0) << 8) | GET_BYTE(to_push, 1)) * 0.1 - 3200U;
```

Reply:
Ah, I've just realised I meant to change this. In the Tesla safety code, they convert the value to the same units as what is used in the steering command. For the Ford, we will have to convert from degrees to radians I think.

```
Steering angle: (0.1 * val) - 3200 in deg
Steering command: (0.0005 * val) - 0.5 in radians
```

Reply:
Switched to `StePinComp_An_Est` signal, as it accounts for the pinion angle being initialised to zero at the start of the drive (no matter how the wheel was oriented). I've adjusted the scale and offset for this too.